doc/proxy-clients.rst

```diff
@@ -296,6 +302,66 @@ And adapt your bootstrapping code to use the cache kernel::
     $response = $cacheKernel->handle($request);
     ...
 
+Cloudflare Client
+~~~~~~~~~~~~~~~~~
+
+.. note::
+
+    Cloudflare does not cache HTML pages by default. To cache them, you need to
+    enable `custom caching with page rules`_ in the cloudflare administration
+    interface.
+
+    The cloudflare client does invalidation requests with the `Cloudflare Purge API`_.
+
+The `Cloudflare`_ client sends HTTP requests with the ``HttpDispatcher``.
+Create the dispatcher as explained :ref:`above <HTTP client configuration>`.
+Set the `server` list to the Cloudflare API `['https://api.cloudflare.com']`.
+You do not need to configure a base URI.
+
+Then create the Cloudflare client with the dispatcher. You also need to pass
+the following options:
+
+* ``authentication_token``: User API token for authentication against
+  Cloudflare APIs, requires `Zone.Cache` Purge permissions.
+* ``zone_identifier``: Identifier for the Cloudflare zone you want to purge the
+  cache for (see below how to obtain the identifier for your domain).
+
+A full example could look like this::
+
+    use FOS\HttpCache\ProxyClient\Cloudflare;
+    use FOS\HttpCache\ProxyClient\HttpDispatcher;
+
+    $options = [
+        'authentication_token' => '<user-authentication-token>',
+        'zone_identifier' => '<my-zone-identifier>',
+    ];
+
+    $httpDispatcher = new HttpDispatcher(['https://api.cloudflare.com']);
+    $cloudflare = new Cloudflare($httpDispatcher, $options);
+
+.. note::
+
+    Cloudflare supports different cache purge methods depending on your account.
+    All Cloudflare accounts support purging the cache by URL and clearing all
+    cache items. You need a `Cloudflare Enterprise`_ account to purge by cache
+    tags.
+
+    When purging the cache by URL see the `Cloudflare Purge by URL`_ docs for
+    information about how Cloudflare purges by URL and what headers you can
+    pass to a :doc:`invalidatePath() <cache-invalidator>` request to clear the
+    cache correctly.
```

Comment:
@simonrjones do you have a code example how purging would look? maybe one with the simple case, and one with special cloudflare headers?

Reply:
I've only tested this with plain URLs (no headers), but an example:

```
An example of clearing the cache by URL::

    $cacheInvalidator->invalidatePath('https://example.com/path')->flush();
```


I think you'd also need to add on line 329:

```
    $cacheInvalidator = new CacheInvalidator($cloudflare);
```

Reply:
thanks. yeah adding $cacheInvalidator would make it complete.

i think i will change the doc where i talk about the baseUri to explicitly say that baseUri does not work with the cloudflare client. (other clients send single requests to the url that should be purged, so the dispatcher can add the domain to the url and invalidation could be just '/path'. in symfony you could then generate just the path of the url, which can be handy e.g. in console context where the domain name might not be known. if we would want the same with the cloudflare client, we would need to pass the baseUri to the cloudflare client and have that logic there. its rather involved though: https://github.com/FriendsOfSymfony/FOSHttpCache/blob/843a10292ba5496b5adb2a143eac5b37ead13680/src/ProxyClient/HttpDispatcher.php#L194-L218 - if we want to share it, we could extract this into a trait that we use in both HttpDispatcher and the Cloudflare client. but we can also do that later, if the need even arises)

Reply:
i updated the documentation. care to have another look if it makes sense like this, before i merge it?

Reply:
Great, all looks good to me 👍